site/topics/testing/testing.rst

```diff
@@ -78,6 +78,89 @@ Writing Tests
+        E.g.
+            square_if_sum(2, 3) -> 25
+
+        :param a: First number
+        :param b: Second number
+        :return: The square of the sum of a and b
+        """
+        c = a + b
+        d = c * c
+        return d
+
+
+    # Tests for square_of_sum function
+    assert 0 == square_of_sum(0, 0)
+    assert 0 == square_of_sum(1, -1)
+    assert 100 == square_of_sum(5, 5)
+    assert 100 == square_of_sum(-5, -5)
+    # To address precision issues, we can look for a sufficiently small difference between the expected and actual
+    assert 0.001 > abs(square_of_sum(2.2, 2.2) - 19.36)
+
+
+* In the above example, the ``square_of_sum`` function is tested a number of times under different input cases
+* Take care to notice that the cases are not just testing different arbitrary input values, but the input values are trying to capture broader cases
+
+    * What happens when the input is zero?
+    * The input has a positive and negative?
+    * There are two positives?
+    * The input is all negative?
+    * What happens when we have floating point numbers?
+
+* We look to capture the broad cases as it is not reasonable to test all possible inputs
+* Further, it's not necessary to test all possible cases
+
+    * If we test ``square_of_sum(5, 5)``, it's reasonable to assume that ``square_of_sum(6, 6)`` would also be fine
+
+
+.. note::
+
+    Notice how the last test looks a little different --- ``assert 0.001 > abs(square_of_sum(2.2, 2.2) - 19.36)``.
+
+    This will be discussed in more detail a little later in the course, but briefly, computers are not great with
+    floating point numbers.
+
+        * What comes after the integer :math:`1`? That's easy, it's :math:`2`.
+        * What comes after the floating point number :math`1.0`? Is it :math`1.1`? Or :math`1.01`? Maybe :math`1.00001`?
+
+    If I run ``square_of_sum(2.2, 2.2)``, the correct answer is ``19.36``, but Python will say ``19.360000000000003``
+    due to the floating point number issue.
+
+    The simple way to address this issue is to check that the *absolute difference* between the expected answer and
+    the function's result is less than some threshold. In the above example, the absolute difference between the
+    correct answer and what Python says is ``0.000000000000003``, which is a tiny difference; the numbers are nearly
+    identical. I chose a threshold of ``0.001``, so if the absolute difference between the expected and actual result is
+    less than that threshold, I will consider that a passed test.
+
+    The choice of the threshold will depend on the situation. Above I could have picked a much smaller number and the
+    test would have passed. But imagine you are a chemist using instruments that can measure to the nearest milliliter.
+    There would be no sense testing beyond a difference of ``0.001`` liters since you cannot get beyond that precision
+    in real life with those instruments.
```

Comment:
This exists in the floating point representation, it's called epsilon. In python, you get it from `sys.float_info.epsilon`. This is the typical offset value you use when testing float point equality.  
Probably not worth mentioning at this point, kinda detracts from the testing point. Maybe worth a bullet point later in the course.